Add action update

kl52752 opened the pull request.

kl52752 commented:
/asign @bowei 
/cc @yazanri

kl52752 commented:
/assign @bowei

bowei commented:
Commit message does not contain much information about this change. Please repush with a detailed commit message on what this change does.

kl52752 commented:
> Commit message does not contain much information about this change. Please repush with a detailed commit message on what this change does.

updated.
I added changes in health check to test this in cloud. Do you want me to split those changes?

bowei commented:
Make them separate commits.

kl52752 commented:
> Make them separate commits.

done

kl52752 commented:
/cc @mag-kol

kl52752 commented:
/cc tangenti

kl52752 commented:
/cc tangenti

tangenti commented on pkg/cloud/rgraph/rnode/action_delete.go:
```diff
@@ -79,7 +79,7 @@ func (a *genericDeleteAction[GA, Alpha, Beta]) Run(
 	}
 
-	a.end = a.start
+	a.end = time.Now()
```
I'd suggest moving this to the top of the function as a `defer` statement

bowei replied:
Do *not* use defer for setting variables. It leads to extremely hard to find bugs.

tangenti commented on pkg/cloud/rgraph/rnode/action_update.go:
```diff
@@ -18,34 +18,95 @@ package rnode
 	// the ambient signal for existance from Update op collides with a
 	// reference to it?
-	return nil // TODO: finish me
+	if got.State() != NodeExists && want.State() != NodeExists {
+		return nil, fmt.Errorf("node for update does not exist")
+	}
+	if want.Ownership() != OwnershipManaged {
+		return nil, fmt.Errorf("node for update is not managed")
+	}
```
So no-ops are treated as an error?

bowei replied:
I'm not sure we want to check here for Ownership, as this happens in a layer above.

tangenti commented on pkg/cloud/rgraph/rnode/action_create.go:
Why the preconditions check here does not check the existence and whether it's managed?

bowei commented on pkg/cloud/rgraph/rnode/action_create.go:
```diff
@@ -79,7 +79,7 @@ func (a *genericCreateAction[GA, Alpha, Beta]) Run(
 ) (exec.EventList, error) {
 	a.start = time.Now()
 	err := a.ops.CreateFuncs(c).Do(ctx, a.id, a.resource)
-	a.end = a.start
```
Create a PR against the main branch to fix this.

bowei commented on pkg/cloud/rgraph/rnode/action_update.go:
```diff
@@ -18,34 +18,95 @@ package rnode
 }
 
 func (a *genericUpdateAction[GA, Alpha, Beta]) DryRun() exec.EventList {
+	// Update Action does not need to notify others on complete.
```
This is not entirely true if references change?

kl52752 commented:
cc/ @mag-kol

bowei commented on pkg/cloud/rgraph/rnode/action_update.go:
```diff
@@ -18,34 +18,122 @@ package rnode
+	wantOutRefs := want.OutRefs()
+	gotOutRefs := got.OutRefs()
+
+	diff := make(map[meta.Key]ResourceRef)
```
do you use the value or can it be struct{}

kl52752 replied:
done

bowei commented on pkg/cloud/rgraph/rnode/action_update.go:
```diff
@@ -18,34 +18,122 @@ package rnode
+		diff[*r.To.Key] = r
+	}
+
+	for _, wantRef := range wantOutRefs {
```
What if we made the map of the wantRefs and then constructed the droprefevents directly?

Seems like the set of dropped = got - want.

then the logic would be 

```
wantMap := ...
for _, gotRef := range gotOutRefs {
  if gotRef not in wantMap { append }
}
```

bowei replied:
would not have to delete map entries and loop again.

kl52752 replied:
done

bowei commented on pkg/cloud/rgraph/rnode/action_update.go:
```diff
@@ -18,34 +18,122 @@ package rnode
+	}
+
+	var events exec.EventList
+	// Drop reference for resources that does not exists in want Node.
```
This comment should be around the whole code that computes the diff

kl52752 replied:
done

kl52752 commented:
/retest
